### DISCUSSION TITLE
basic-types/numbers

### DISCUSSION THREAD

**Discussion by sunface** [sunface/rust-by-practice#166] [Book Comments]
https://practice.rs/basic-types/numbers.html 

 Learn Rust with Example, Exercise and real Practice, written with ❤️ by https://course.rs team

**Comment by ZhangBin-Rs** [comment]
大佬位操作没有讲呢

**Reply by sunface** [reply]
你是说 course.rs 中没有讲嘛？

**Reply by ZhangBin-Rs** [reply]
是的，基本类型那一章没看到有。

**Reply by sunface** [reply]
好的～～

**Comment by EdValon** [comment]
第6题的要求是不是说法有点问题，我感觉你想要的是断言里为true，而不是代码工作，我理解断言哪怕为false也是工作的。被这个要求卡了一段时间，我一开始以为我理解的i32类型可以用"=="比较判断是错误的

**Reply by sunface** [reply]
我看看哈～

**Reply by suiang** [reply]
断言 == 改为 != 就通过了😂

**Reply by Leopold-Fitz-AI** [reply]
其实还可以把那个数字算出来，改成1597😄

**Comment by sweet2honey** [comment]
sun大你好，第二题中的`as`操作，貌似在`2.2.1`章节里面没有讲噢

**Reply by sunface** [reply]
收到～～

**Reply by sweet2honey** [reply]
还有11题中，截止到这一章节，其实还没有引入布尔值的概念，短路运算也没有提过，下一节也简单带过了；不过有编程背景所以还好~
然后就是位运算的习题那里，发现`0b0011u32`这种用法（或者`0b0011_u32`），很nice！可以在教程中提一下这种指定类型的用法

**Reply by sunface** [reply]
可以提 PR 哈哈

**Comment by Rackar** [comment]
```
for c in 'a'..='z' {
        println!("{:?}",c as u8);
    }
```
说出来大家可能不信，这是我试了几次不行，Copilot教我的

**Reply by sunface** [reply]
还能这样？哈哈

**Reply by roushanyyzz** [reply]
请问{:?}是什么意思呢？

**Reply by ZeyuanGuo** [reply]
直接
```
 for c in 'a'..='z' {
        println!("{}",c as i32);
    }
```
hhh

**Reply by 844189873** [reply]
看一下char模块里面convert.rs的源码，里面有impl From<char> for u32{...}，于是我像下面这样试了一下，果然可以：

for c in 'a'..='z' {
        println!("{}",u32::from(c));
    }

**Comment by fireindark707** [comment]
完成，11/11

**Comment by RomanSeven7** [comment]
c as u8 真是秀  我还在想 字符转为字符串 在转int呢

**Reply by michael1735** [reply]
可以用b'a'..=b'z'

**Comment by roushanyyzz** [comment]
第八题第二种解法有吗？
我只知道一种
```rust
fn main() {
    assert!(0.1_f32+0.2_f32==0.3_f32);
}

```

**Reply by roushanyyzz** [reply]
额，后面看到答案了，我还是贴一下
```rust
fn main() {
    assert!((0.1_f64+ 0.2 - 0.3).abs() < 0.001);
}
```

**Reply by tkit1994** [reply]
```rust
fn main() {
    assert!(0.1_f32+0.2_f32==0.3_f32);
}
```
为什么f64不可以，f32可以。

**Reply by jackyliu16** [reply]
非常简单，精度问题，圣经里面也提到过这个内容，要尽量避免在浮点数中使用等于运算。

**Reply by ZeyuanGuo** [reply]
可以只写一个f32（？）
`0.1+0.2==0.3 as f32`
应该是为了不转换类型，会把所有的都推断成f32
但是赌f32不出精度问题也不太好

**Reply by Passer-by** [reply]
assert!(0.1+0.2!=0.3); 
思路要打开（狗头保命

**Reply by xyanyue** [reply]
`assert!(((0.1+0.2)*10_f64).round()/10_f64==0.3);`

实际编程中，比较浮点数，将其转为int是比较式常用方法。答案的方法只适用这道题
```rust
// 比较ab浮点数在第几位之前是否相等
fn eq_float64(a:f64,b:f64,position:i32) -> bool {
    let factor = 10_f64.powi(position);
    (a * factor).trunc() == (b * factor).trunc()
}

```

**Comment by MikeMMS** [comment]
第6题, 把==改成!=就行了~

**Reply by MikeMMS** [reply]
同理第8题, 两种解法一种改成!=, 另一种改成>

**Reply by MikeMMS** [reply]
同理第9题第一问也是改成!=

**Reply by mobeicanyue** [reply]
你这么算 算了个啥呀…
作者出的==本来就是错的答案，让你把等式右边 改成对的 程序才能跑通。
你这另辟蹊径，直接告诉编译器是错的能跑通就行？那练习意义何在

**Comment by jackyliu16** [comment]
想问下，这个题目的**目的**是不是不在于对应的对错，而在于对于知识点的掌握，就比如说第五题
只要我能发现产生panic的原因在于整数溢出，是不是就可以了？
有很多方法可以使结果成功，就比如说我之前看到的循环补码运算（前面的溢出放到后面）也是个操作方法
【淦，真的忘记在哪里看到的了】

**Comment by Lightblues** [comment]
请问 `0b0011u32` 最后为什么要加 u32?

**Reply by HydroCarbon** [reply]
表示是无符号的 32 位整形数

**Comment by wd2010** [comment]
大佬，请问 `println!("0011 AND 0101 is {:04b}", 0b0011u32 & 0b0101);` 中 `{:04b}`是什么意思？

**Reply by fbbin** [reply]
{:04b} 0 代表左侧补0， 4 代表宽度是4，b代表转成二进制

**Reply by HydroCarbon** [reply]
对输出的格式化，长度不足 4 位的进行高位补 0 打到四位，b 是 binary 的简称，表示格式为 2 进制数。

**Comment by MatRunner** [comment]
No.9 序列
`1..=5`和`1..6`是一样的吗？

**Reply by killdayu** [reply]
对的，都是输出1,2,3,4,5

**Reply by tayuqitan** [reply]
结果一样但是类型不一样：
(1..5) 的时候类型的Range
(1..=5) 的时候类型的RangeInclusive

**Reply by zhy-code** [reply]
我也想问这个问题，看到回复了

**Comment by axzed** [comment]
mark~

**Comment by worldstearly** [comment]
我真的是遇到了神一样的编译器，都怪我太菜了😀
```
error: literal out of range for `i8`
  --> test4.rs:44:25
   |
44 |     assert_eq!(i8::MAX, 255); 
   |                         ^^^
   |
   = note: `#[deny(overflowing_literals)]` on by default
   = note: the literal `255` does not fit into the type `i8` whose range is `-128..=127`
   = help: consider using the type `u8` instead
```

**Reply by saadaltafofficial** [reply]
u8 = 0 to 255
i8 =-128 to 128

**Comment by twandylue** [comment]
Done~

**Comment by xrr2016** [comment]
Done~

**Comment by XuanZhou-mars** [comment]
第10题第二个 为啥不能写 1..6 而必须写 1..=5 哇，有没有大佬解释一下

**Reply by Gray-min** [reply]
后面类型写的是1到5，虽然从输出上面来说1..6和1..=5都是1到5这几个数字

**Reply by Non-PlayerC** [reply]
Range 和 RangeInclusive 都是 Rust 的区间类型。它们表示一个左闭右开的区间（Range）或者是一个左闭右闭的区间（RangeInclusive），都包含了一个起点和一个终点，用于表示一段连续的值的范围。

其中 Range 使用的是 .. 符号来表示区间，左闭右开；而 RangeInclusive 使用的是 ..= 符号来表示区间，左闭右闭。

RangeInclusive::new 是一个构造函数，可以用来创建一个左闭右闭的区间对象。

**Comment by cddsgtc** [comment]
pass

**Comment by imsellbaox** [comment]
maker

**Comment by zhe184324** [comment]
求大神指点，为什么我的第10题写1..6会报错呀。。

assert_eq!((1..6), RangeInclusive::new(1, 5));
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::ops::Range`, found struct `RangeInclusive`
  |

**Reply by zhe184324** [reply]
（1..=5）是ok的。

**Reply by tayuqitan** [reply]
(1..5) 的时候类型的Range
(1..=5) 的时候类型的RangeInclusive
需要类型匹配

**Comment by lam2018** [comment]
第9道题一度认为是要打印出 97-122这个数学公式，直到我看到了答案。

**Comment by Non-PlayerC** [comment]
😊

**Comment by XU0618** [comment]
lala~pass

**Comment by tutdahai** [comment]
打卡

**Comment by Chenleely** [comment]
done

**Comment by caioo0** [comment]
1.

use std::fmt::{Debug};


fn type_v<T>(t:T) where T:Debug {
    println!("{:?},{}",t,core::any::type_name::<T>());
 }
// 移除某个部分让代码工作
fn main() {
    let x: i32 = 5;
    let mut y: i32 = 5;

    y = x;
    
    let z = 10; // 这里 z 的类型是? 
    type_v(z);
}

**Comment by ZeyuanGuo** [comment]
第四题这里这样写好像是把未位操作先推断成u8了，有便捷的写法解决这个推断吗
```
    assert_eq!(i8::MAX, (1<<7)-1)); 
    assert_eq!(u8::MAX, (1<<8)-1)); 
```

**Comment by jackrobet** [comment]
mark

**Comment by Ezreal09** [comment]
这些数字你们是怎么记住的

**Comment by Muzych** [comment]
Mark 好难啊rust

**Comment by ledunguit** [comment]
Got it!

**Comment by Zhaoyikaiii** [comment]
fn main() {
    // 方式一：使用浮点数的近似相等性
    // 为了解决这个问题，可以通过比较精度来判断两个浮点数是否相等
    let x = 0.1 + 0.2;
    assert!(x - 0.3 < 0.00001);

    // 方式二
    assert!(0.1_f32 + 0.2_f32 == 0.3_f32)
    //_f32 后缀表示这个数字的类型是 32 位浮点数，而不是默认的 64 位。
    // 32 位浮点数是单精度浮点数，它们是 f32 类型的值。
    // 64 位浮点数是双精度浮点数，它们是 f64 类型的值。
    // _f32和as区别是什么？
    // 1. _f32是rust语法，as是强制类型转换
}

**Comment by Zhaoyikaiii** [comment]
fn main() {
    let mut sum = 0;
    for i in -3..2 {
        sum += i
    }// -3 + -2 + -1 + 0 + 1 注意：不包含2
    assert!(sum == -5);
    for c in 'a'..='z' {
        print!("{}", c);
    }
}

**Comment by Zhaoyikaiii** [comment]
// 填空
use std::ops::{Range, RangeInclusive};
fn main() {
    assert_eq!((1..__), Range{ start: 1, end: 5 });
    assert_eq!((1..__), RangeInclusive::new(1, 5));
}

**Comment by KGIM** [comment]
第5题是不是在2.2.1之前没有讲到过这个函数？

**Comment by xiang-noodle** [comment]
done--2023/12/9

**Comment by yjshi2015** [comment]
0b0110_u32 这种写法泰酷辣

**Comment by thiru-01** [comment]
done

**Comment by Sahil-Shadwal** [comment]
Done!

**Comment by AcidVienna** [comment]
I was here :)

**Comment by YerboyMike** [comment]
cool beans

**Comment by zhengkezhou1** [comment]
done

**Comment by vickyone0** [comment]
done  batman

**Comment by EmptyBlueBox** [comment]
checked!

**Comment by TDAkory** [comment]
mark finish

**Comment by Prajwalraut29** [comment]
done

**Comment by vickyone0** [comment]
anyone , how to limit the decimal in the problem 8 
fn main() {
    assert!(0.1+0.2==0.3);

    println!("Success!");
}
 with least modification

**Reply by johnviggo** [reply]
You can use f32, since it is less accurate than f64 which is used here.

assert!(0.1 as f32 +0.2 as f32 ==0.3 as f32);

**Comment by zhengkezhou1** [comment]
vickyone0 ***@***.***> 于 2024年4月4日周四 17:28写道：

**Comment by ItzAus6on** [comment]
done

**Comment by erhszo** [comment]
Done

**Comment by rishabhnair12345** [comment]
done

**Comment by C-RONALDO7V** [comment]
mark

**Comment by c-shankar** [comment]
done

**Reply by anousonefs** [reply]
ແລ້ວໆ

**Comment by sjunepark** [comment]
Done

**Comment by jidalii** [comment]
For Q11, can we do `assert!(9.6_f32 / 3.2_f32 == 3.0_f32);`

**Comment by lichenbuliren** [comment]
done

**Comment by hoBabu1** [comment]
completed

**Comment by jieqiyue** [comment]
第八题，
```
fn main() {
    assert!((0.1_f64+ 0.2 - 0.3).abs() < 0.001);
}
为什么要在0.1后面加上f64啊，我谷歌了一下，说是因为assert不能分辨计算后的类型，所以要显示的指定类型。
我的问题是：rust语言默认是f64啊，为什么这里还需要指定类型呢？如果不指定类型的话，不也是f64吗？为什么指定了类型之后就可以运行了啊？
好心人帮我看看

**Reply by craftopc** [reply]
对浮点数使用 abs() 必须要显式的指定类型

```bash
❯ cargo run
   Compiling variables v0.1.0 (/home/craftopc/code/rust/variables)
error[E0689]: can't call method `abs` on ambiguous numeric type `{float}`
 --> src/main.rs:7:9
  |
7 |     1.0.abs()
  |         ^^^
  |
help: you must specify a concrete type for this numeric value, like `f32`
  |
7 |     1.0_f32.abs()
  |     ~~~~~~~

For more information about this error, try `rustc --explain E0689`.
error: could not compile `variables` (bin "variables") due to 1 previous error
```

**Comment by yomanthunder** [comment]
learned a lot , thanks

**Comment by lihejia** [comment]
完成，11/11

**Comment by zhengkezhou1** [comment]
day1

**Comment by Paul-from-NJTech** [comment]
done

**Comment by MahajanSamiksha** [comment]
done

**Comment by Abhinav-Singh10** [comment]
😁

**Comment by UnclePhil1** [comment]
Completed

**Comment by chttiu** [comment]
标记完成

**Comment by Tenax-599** [comment]
十一题第五个还可以
assert!(9.6f32 / 3.2f32 == 3.0);

**Comment by mdShakilHossainNsu2018** [comment]
done

**Comment by in-people** [comment]
fn main() {
    let v = 1_024 + 0xff + 0o77 + 0b1111_1111;
    assert!(v == 1597);
}
// 1024 + 255 + 63 + 255

**Comment by eyenov** [comment]
完成

**Comment by obhor** [comment]
omkar was here

**Comment by ebdurrehm** [comment]
done
